File: lib/core/window.ml
open Core
module Wire = Gpuio_protocol.Window_wire
module Chrome = Wire.Chrome
module Backend = Wire.Backend
module Capabilities = Wire.Capabilities
module Snapshot = Wire.Snapshot
module Command = Wire.Command
module Error = Wire.Error

module Config = struct
  type t = Wire.Config.t

  let create
        ?(focus = true)
        ?(chrome = Chrome.Standard)
        ?(resizable = true)
        ~title
        ~width
        ~height
        ()
    =
    if Wire.valid_title title && Wire.valid_size width height
    then Ok { Wire.Config.title; width; height; focus; chrome; resizable }
    else Or_error.error_string "invalid window title or logical size"
  ;;

  module Expert = struct
    let to_wire t = t
  end
end

module Close_reason = struct
  type t =
    | Window_close
    | Application_quit
  [@@deriving equal, sexp_of]
end

module Close_decision = struct
  type t =
    | Allow
    | Keep_open
  [@@deriving equal, sexp_of]
end

File: rust/protocol/src/window.rs
//! Window requests use logical pixels. Acknowledgements observe compositor state;
//! they do not promise a requested resize/fullscreen transition has completed.
use binprot::macros::BinProtWrite;
#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum Chrome {
    Standard,
    Hidden,
}
#[derive(Clone, Debug, PartialEq, BinProtWrite)]
pub struct Config {
    pub title: String,
    pub width: f64,
    pub height: f64,
    pub focus: bool,
    pub chrome: Chrome,
    pub resizable: bool,
}
pub fn valid_title(title: &str) -> bool {
    !title.is_empty() && title.len() <= 4096 && !title.contains('\0')
}
pub fn valid_size(width: f64, height: f64) -> bool {
    width.is_finite()
        && height.is_finite()
        && (1.0..=16384.0).contains(&width)
        && (1.0..=16384.0).contains(&height)
}
impl Config {
    pub fn is_valid(&self) -> bool {
        valid_title(&self.title) && valid_size(self.width, self.height)
    }
}
#[derive(Clone, Debug, PartialEq, BinProtWrite)]
pub enum Command {
    Observe,
    SetTitle(String),
    Resize(f64, f64),
    Activate,
    Zoom,
    ToggleFullscreen,
    SetEdited(bool),
}
impl Command {
    pub fn is_valid(&self) -> bool {
        match self {
            Self::SetTitle(title) => valid_title(title),
            Self::Resize(w, h) => valid_size(*w, *h),
            _ => true,
        }
    }
}
#[derive(Clone, Debug, PartialEq, BinProtWrite)]
pub struct Snapshot {
    pub title: String,
    pub x: f64,
    pub y: f64,
    pub width: f64,
    pub height: f64,
    pub content_width: f64,
    pub content_height: f64,
    pub active: bool,
    pub fullscreen: bool,
    pub maximized: bool,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum Backend {
    Macos,
    Wayland,
    X11,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub struct Capabilities {
    pub backend: Backend,
    pub native_quit_decision: bool,
}
#[derive(Clone, Copy, Debug, PartialEq, Eq, BinProtWrite)]
pub enum Error {
    Closed,
    NotReady,
    Busy,
    InvalidRequest,
    NativeFailure,
}
#[derive(Clone, Debug, PartialEq, BinProtWrite)]
pub enum Response {
    Observed(Snapshot),
    Failed(Error),
}
